keps/sig-apps/0003-kubernetes-application-api.md

```diff
@@ -0,0 +1,1120 @@
+no longer installed.
+
+For all objects other than Applications, when an object is updated, the Application Controller will do the following.
+
+1. [Retrieve the Application for the object](#retrieving-applications).
+1. If the Application's `Selector` does not match the object's labels, the Application Controller will update the 
+Applications `Status` to indicate that component is no longer installed and remove any `OwnerReference` from the 
+Application to the object.
+
+#### Health Check Pods
+If the Application Controller observes the presence of an ApplicationHealthCheck object in the `Spec` of an 
+Application, it will do the following.
+
+1. Generate a unique name for a Pod based on the hash of the `Template` field of the HealthCheck. The name generation 
+method should be analogous to that of ReplicaSet and Deployment.
+1. Create a new Pod based on the `Template`.
+1. The Application Controller shall delete and recreate HealthCheck Pods upon mutation of `Spec.HealthCheck` field of 
+the application.
+
+#### Health Status Reporting.
+The Application Controller will do the following to update the `Stauts.Ready` ConditionStatus of an Application object.
+
+1. If the Application Controller observes an Application object with a nil `Spec.HealthCheck`, it shall set the 
+`Status.Ready` field of the Application object to `UNKNOWN`.
+1. If the Application Controller observes a Pod created from an ApplicationHealthCheck it shall 
+[retrieve the Application for the object](#retrieving-applications).
+   1. If the Pod has no Ready Condition, the Application Controller shall set the `Status.Ready` field of the Application 
+   object to `UKNOWN`.
+   1. If the Pod has a Ready Condition, the Application Controller shall set the `Status.Ready` field of the Application 
+   object to the Status of the Pod's Ready Condition.
+
+#### Generation Observation
+Each time the Application Controller observes the creation or mutation of an Application object, it will set the 
+`Status.ObservedGeneration` to the `Generation` of the Application. As monotonicity of the object's `Generation` is 
+ensured by the API Server, the `Satatus.ObservedGeneration` can be used to determine if the Application Controller 
+has observed a mutation an Application object.
+
+#### Adopting Existing Components
+The design of [Garbage Collection](#garabage-collection) allows existing objects to be aggregated into an Application. 
+In order to achieve this users should do the following.
+
+1. Create an Application object that contains the objects that will be adopted in the Application's `Spec.Components`.
+1. Annotate the objects with a `kubernetes.io/application` label that references the Application.
+
+The Application Controller will allow the newly created Application to adopt the existing object. By design, this 
+workflow should be easily achievable programmatically. Existing tools may wish to implement a command that creates 
+an Application in this manner.
```

Comment:
See prior point about preserving PVCs. Preserved objects would become adoptable, which is probably a desired behavior in some cases. E.g. if I destroy and re-created a database, it is at least somewhat likely that if I said I didn't want the PVC deleted, I intended for it to be re-adopted when creating a new instance (with the same name) as last time.

Reply:
We will start with name based GC.